syne_tune/optimizer/schedulers/searchers/searcher.py

```diff
@@ -551,17 +568,139 @@ def __init__(
+            # num_samples is ignored for categorical hps.
+            if isinstance(hp_range, Categorical) or isinstance(hp_range, FiniteRange):
+                if hp in self.num_samples:
+                    logger.info(
+                        'number of samples for categorical variable "{}" is ignored.'.format(
+                            hp
+                        )
+                    )
+
+    def _remove_duplicate_candidates(self, all_candidates) -> List[dict]:
+        """
+        Excludes _points_to_evaluate from all_candidates list, so that no duplicate
+        candidates exist.
+        Args:
+            all_candidates:  all candidates on grid generated by _generate_all_candidates_on_grid()
+
+        Returns:
+            de-duplicated list of candidates on grid
+        """
+        excl_list = ExclusionList.empty_list(self._hp_ranges)
+        for candidate in self._points_to_evaluate:
+            excl_list.add(candidate)
+
+        remaining_candidates = []
+        for candidate in all_candidates:
+            if not excl_list.contains(candidate):
+                remaining_candidates.append(candidate)
+                excl_list.add(candidate)
+
+        self._grid_candidates = remaining_candidates
+        self._candidate_index = 0
+        return
+
+    def _generate_all_candidates_on_grid(self) -> List[dict]:
+        """
+        Generates all configurations to be evaluated by placing a regular, equally spaced grid over the configuration space
+        Returns:
+            The set of all configurations on grid.
+        """
+        hp_keys = []
+        hp_values = []
+        ## adding categorical, finiteRange, scalar parameters
+        for hp, hp_range in reversed(list(self.config_space.items())):
+            if isinstance(hp_range, Float) or isinstance(hp_range, Integer):
+                continue
+            if isinstance(hp_range, Categorical):
+                values = hp_range.categories
+            elif isinstance(hp_range, FiniteRange):
+                values = hp_range.values
+            elif not isinstance(hp_range, Domain):
+                values = [hp_range]
+            hp_keys.append(hp)
+            hp_values.append(values)
+
+        ## adding float, integer parameters
+        for hpr in self._hp_ranges._hp_ranges:
+            if hpr.name not in hp_keys:
+                _hpr_nsamples = self.num_samples[hpr.name]
+                _normalized_points = [
+                    (idx + 0.5) / _hpr_nsamples for idx in range(_hpr_nsamples)
+                ]
+                _hpr_points = [
+                    hpr.from_ndarray(np.array([point])) for point in _normalized_points
+                ]
+                _hpr_points = list(set(_hpr_points))
+                hp_keys.append(hpr.name)
+                hp_values.append(_hpr_points)
+
+        self.hp_keys = hp_keys
+        self.hp_values_combinations = list(product(*hp_values))
```

Comment:
Why keep this in `self.hp_values_combinations` if you do not use it afterwards?

I'd be most happy with this, to be honest:
- Keep (and use) `self.hp_keys` and `self.hp_values_combinations` here
- Forget about `all_candidates_on_grid` and also about `_remove_duplicate_candidates`. As I said, you can
   assume that `self.hp_values_combinations` does not contain duplicates, and returning possibly a config or so 2x is not a problem
- Convert the tuple to dict only in `get_config`

Reply:
I think you mean to generate it in `_next_candidate_on_grid()`

Reply:
Assuming that the unrolled list of all grid configs is very large, I'd recommend not storing dict, but tuple, and wrap tuple into grid only in `get_config`. I also recommend not doing the check for duplicates. Will all make the code simopler as well.

Reply:
According to our conversation, we decided to remove de-duplication logic from gridsearch (due to it being expensive) and allow gridSearcher to return duplicate configuration (this happens when points_to_evaluate is not None) As the result one of previous unit tests i.e.`test_store_and_restore_state_with_initial_config()` test in tst/schedulers/test_grid_searcher.py fails.

Reply:
basically, in the current implementation, if init_config (points_to_evaluate) is given, gridSearcher still produces those points and it leads to duplicate configurations.

Reply:
If this test is only testing this feature (no duplicates if points_to_evaluate is given), I'd remove it. If it tests something else, I'd just choose points_to_evaluate to not contain grid points.

Reply:
to double check: we expect initial_points to be in the config_space, correct? As an example, the two following is NOT a proper setup: 
config_space = {
        "char_attr": choice(["a", "b"]),
        "int_attr": choice([1, 2]),
    }
inital_config = [
        {"char_attr": "c", "int_attr": 1},
        {"char_attr": "c", "int_attr": 2},
    ]

Reply:
`test_store_and_restore_state_with_initial_config()` also tests to proper restoring of gridsearcher's state when initial_points is given. So I modified the test to remove the test for de-duplication, and would like to keep it.